### ISSUE TITLE
Stack overflow on `WITH cte AS (SELECT * FROM cte) SELECT * FROM cte`

### ISSUE THREAD

**Issue by jwills** [cwida/duckdb#1050]
Came across this when I was working on #1048: I'm trying to make DuckDB work like Postgres, where it's okay for CTEs inside of subqueries to override the name of a CTE in the parent binding scope. For example, this query works correctly on Postgres:

```
create table integers (i int);
insert into integers VALUES (3), (4), (5);

WITH cte AS (WITH cte AS (SELECT sum(i) si FROM integers) SELECT * FROM cte) SELECT * FROM cte;
 si 
----
 12
(1 row)
```

On DuckDB, trying to run that sequence (or the even simpler one I put in the title of this issue) against `master` will cause a stack overflow `in duckdb::Binder::GenerateTableIndex() binder.cpp:174`.

I feel like the right fix is to make a CTE unable to refer to its own name as a term in the subquery binding (although for my use case to work correctly, it *should* be able to refer to an identically named CTE defined in the parent binding), but I'm not sure of the best place to handle that case in the code.

**Comment by Mytherin** [comment]
Thanks, nice catch! Can confirm this causes a stack overflow. It indeed seems like the CTE references itself, which should not be possible in a non-recursive case. I will have a look to see if I can come up with an elegant fix here.

**Comment by jwills** [comment]
Fixed by PR #1048
